clevertap-core/src/main/java/com/clevertap/android/sdk/inapp/customtemplates/CustomTemplate.kt

```diff
@@ -0,0 +1,271 @@
+     * structure. They are treated the same way as the keys provided by [mapArgument]. If a higher-level name (to the left
+     * of a . symbol) matches a map argument's name it is treated the same as if the argument was part of the map itself.
+     *
+     * For example, the following code snippets define identical arguments:
+     * ```
+     * builder.mapArgument(
+     *     name = "map",
+     *     value = mapOf (
+     *         "a" to 5,
+     *         "b" to 6
+     *     )
+     * )
+     * ```
+     * and
+     * ```
+     * builder.intArgument("map.a", 5)
+     * builder.intArgument("map.b", 6)
+     * ```
+     */
+    class TemplateBuilder : Builder<TemplateBuilder>(TEMPLATE, allowHierarchicalNames = true, isVisual = true) {
+
+        override val thisRef: TemplateBuilder = this
+
+        /**
+         * Add a map structure to the arguments of the [CustomTemplate]. The [name] should be unique across all
+         * arguments and also all keys in [value] should form unique names across all arguments.
+         *
+         * @param value The map must be non-empty. Values can be of type [Byte], [Short], [Int], [Long], [Float],
+         * [Double], [Boolean], [String] or another [Map]<String, Any> which values can also be of the same types.
+         */
+        fun mapArgument(name: String, value: Map<String, Any>): TemplateBuilder {
+            if (value.isEmpty()) {
+                throw CustomTemplateException("Map argument must not be empty")
+            }
+
+            for (mapEntry in value) {
+                val argValue = mapEntry.value
+                val argName = "$name.${mapEntry.key}"
+
+                @Suppress("UNCHECKED_CAST")
+                when (argValue) {
+                    is Byte -> byteArgument(argName, argValue)
+                    is Short -> shortArgument(argName, argValue)
+                    is Int -> intArgument(argName, argValue)
+                    is Long -> longArgument(argName, argValue)
+                    is Float -> floatArgument(argName, argValue)
+                    is Double -> doubleArgument(argName, argValue)
+                    is Boolean -> booleanArgument(argName, argValue)
+                    is String -> stringArgument(argName, argValue)
+                    is Map<*, *> -> mapArgument(argName, argValue as Map<String, Any>)
+                    else -> throw CustomTemplateException("Unsupported value type ${argValue.javaClass} for argument $argName")
+                }
+            }
+            return this
+        }
+
+        fun actionArgument(name: String): TemplateBuilder {
+            addArgument(name, ACTION, null)
+            return this
+        }
+    }
+
+    sealed class Builder<T : Builder<T>>(
+        private val type: CustomTemplateType,
+        private val allowHierarchicalNames: Boolean,
+        private val isVisual: Boolean
+    ) {
+
+        protected abstract val thisRef: T
+
+        private var templateName: String? = null
```

Comment:
can we have a val in constructor itself? it will save mutability and force client to assign a name

Reply:
`templateName` should be validated. Validating in the constructor would have to either have a default value, which would hide the problematic declaration or throw in the constructor which is a bad design. This is one of the main reasons for us to use a builder in this case.